### ISSUE TITLE
Qwen3-Embedding-0.6B forward-pass divergence (cosine 0.948 vs HF on whitespace/CJK inputs)

### ISSUE THREAD

**Issue by ohdearquant** [ohdearquant/lattice#103]
## Symptom

The HF parity regression test from the `embed-perf-quality` work measures Qwen3-Embedding-0.6B failing 3/5 cases against HF reference, with the worst at cosine **0.9477** (leading-whitespace input). On the "fox" input — where lattice tokens match HF token-for-token — vectors still diverge at cosine **0.9886**, max-abs-diff **1.41e-2**.

BGE, E5, all-MiniLM, and paraphrase-multilingual all reach min cosine ≥ 0.9998 across the same 5 inputs after that work's tokenizer + embed fixes. Qwen3-Embedding is the only outlier.

## Failing inputs

```
[qwen3-0.6b] "The quick brown fox jumps over the lazy dog."   cosine=0.988615  max_diff=1.41e-2
[qwen3-0.6b] "   leading whitespace and    multiple    spaces   "  cosine=0.947711  max_diff=3.47e-2
[qwen3-0.6b] "短い日本語のテストです。"                            cosine=0.989882  max_diff=1.74e-2
```

Passing inputs:
```
[qwen3-0.6b] "Pure Rust transformer inference engine."         cosine=0.997272 PASS
[qwen3-0.6b] "Café résumé naïve façade — Unicode test."        cosine=0.995420 PASS (barely)
```

## Important observation

For the "fox" input, **lattice's BPE tokenizer produces identical IDs to HF** (verified — fixture `qwen3_embedding_0_6b.json` records HF's tokenization, and `audit_tokenizer_parity` cases include this exact input). So the divergence is **not** in the tokenizer — it's in the forward pass between encoded IDs and the final pooled embedding.

Per `lattice/CLAUDE.md` quantitative bounds: bf16 vs f32 PPL delta is < 0.05 (negligible). A vector divergence of 1.4e-2 to 3.5e-2 is structurally larger than bf16 noise and points to a real algorithmic difference, not numerical drift.

## Hypotheses (ordered by historical precedent)

1. **RoPE convention mismatch** — Qwen3-Embedding's RoPE differs from Qwen3-Chat's RoPE in some subtle way (e.g., the per-head dimension, the theta base, or pair convention). The v0.2.3 stride-half vs interleaved RoPE bug had this exact symptom shape.
2. **Attention mask convention** — Qwen3-Embedding uses *bidirectional* attention (it's an encoder-style retrieval model), but lattice may be applying causal mask inherited from Qwen3-Chat.
3. **Last-token pooling extraction** — Qwen3-Embedding pools the *last non-padding token*. If lattice uses `seq_len - 1` for fox (no padding) and `attention_mask.sum() - 1` for the longer inputs, that could explain the mixed pattern.
4. **bf16 vs f32 accumulation in attention** — partial f32 promotion in HF that lattice does in bf16.
5. **Output linear / output normalization mismatch** — Qwen3-Embedding may have a final projection or normalization HF applies that lattice skips (or vice versa).

## Reproducer

```bash
cd /path/to/lattice
cargo test -p lattice-embed --test embed_parity_vs_hf -- qwen --nocapture
# OR run only the failing cases:
cargo test -p lattice-embed --test embed_parity_vs_hf -- qwen --ignored --nocapture
```

The fixture file (`crates/embed/tests/fixtures/embed_parity_v1/qwen3_embedding_0_6b.json`) contains HF-reference embeddings; regenerate via `scripts/gen_embed_parity_goldens.py`.

## Why it's not a ship blocker for v0.2.5

Khive's primary retrieval models are BGE-small + MiniLM-L6 + paraphrase-multilingual-MiniLM-L12, all of which pass parity at cosine ≥ 0.9998. Qwen3-Embedding-0.6B is an optional advanced model. For now the test is `#[ignore]`'d with a link back to this issue; uncomment when the bug is fixed.

## Recommended investigation method

Per `lattice/CLAUDE.md` "Differential Test First": write a Python script that:
1. Loads `Qwen/Qwen3-Embedding-0.6B` via `transformers.AutoModel` with `output_hidden_states=True`
2. Runs the "fox" input and dumps per-block hidden states + final layernorm output + last-token vector pre-L2-norm
3. Has lattice dump the same tensors
4. Compares per-block max-diff to find the first divergent layer

Once the divergent layer is identified, the fix is usually small. See v0.2.3 release notes for the RoPE stride-half precedent — 0.77 PPL → 0.029 PPL via a 1-character fix once the layer was identified.

## Investigation already started

An investigation is running on a branch exploring this divergence (running at the time this issue was filed). If their report lands a smoking gun, link it here.

---
Filed via Claude Code — incident date 2026-05-25 as part of the `embed-perf-quality` work.
